pymsis/utils.py

```diff
@@ -48,8 +54,60 @@ def download_f107_ap() -> None:
         f.write(req.read())
 
 
-def _load_f107_ap_data() -> dict[str, npt.NDArray]:
-    """Load data from disk, if it isn't present go out and download it first."""
+def _refresh_f107_ap_data(last_obs_date: np.datetime64) -> None:
+    """
+    Refresh exising SW_All file after last_obs_date.
+
+    Parameters
+    ----------
+    last_obs_date : datetime of last observed parameter
+    """
+    warnings.warn(
+        f"Refreshing data using partial ap and F10.7 data from {_PARTIAL_F107_AP_URL}"
+    )
+    req = urllib.request.urlopen(_PARTIAL_F107_AP_URL)
+    with _PARTIAL_F107_AP_PATH.open("wb") as f:
+        f.write(req.read())
+
+    # Store all observed dates in existing_data
+    with open(_F107_AP_PATH) as f:
+        reader = csv.DictReader(f)
+        existing_data = [
+            row for row in reader if np.datetime64(row["DATE"]) <= last_obs_date
+        ]
+
+    # Store data new dates in updated_data
+    updated_data = []
+    if os.path.exists(_PARTIAL_F107_AP_PATH):
+        with open(_PARTIAL_F107_AP_PATH) as f:
+            reader = csv.DictReader(f)
+            updated_data = [row for row in reader]
+
+    # Merge existing_data and new_data into all_data
+    all_data = {row["DATE"]: row for row in existing_data}
+    all_data.update({row["DATE"]: row for row in updated_data})
+
+    # Sort all_data by date
+    sorted_data = sorted(all_data.values(), key=lambda x: np.datetime64(x["DATE"]))
+    with open(_F107_AP_PATH, "w", newline="") as f:
+        writer = csv.DictWriter(f, fieldnames=sorted_data[0].keys())
+        writer.writeheader()
+        writer.writerows(sorted_data)
+
+    os.remove(_PARTIAL_F107_AP_PATH)
```

Comment:
Be wary of this because if there is any download error above, this file will be left hanging around.

A few quick thoughts:
* Do we need to save it to disk, or could you store it in a BytesIO object in memory instead?
* Would it make sense to save `_DATA` arrays with `np.savez()` instead of the csv files for easier manipulation? (Then you wouldn't need to mess with the dictreader/writer.
* Honestly, now that I'm thinking about it more, is this just more headache than it is worth and we should just use your logic below and re-download the full file for the simplest case. More bandwidth, but at least we are getting the full file built for us rather than building it from bits and pieces.